subprojects/core/src/main/java/org/gradle/api/internal/file/DefaultFileOperations.java

```diff
@@ -156,26 +163,50 @@ public ConfigurableFileTree fileTree(Map<String, ?> args) {
 
     @Override
     public FileTreeInternal zipTree(Object zipPath) {
-        return new FileTreeAdapter(new ZipFileTree(file(zipPath), getExpandDir(), fileSystem, directoryFileTreeFactory, fileHasher), patternSetFactory);
+        Provider<File> fileProvider = asFileProvider(zipPath);
+        return new FileTreeAdapter(new ZipFileTree(fileProvider, getExpandDir(), fileSystem, directoryFileTreeFactory, fileHasher), patternSetFactory);
     }
 
     @Override
     public FileTreeInternal tarTree(Object tarPath) {
-        File tarFile = null;
-        ReadableResourceInternal resource;
-        if (tarPath instanceof ReadableResourceInternal) {
-            resource = (ReadableResourceInternal) tarPath;
-        } else if (tarPath instanceof ReadableResource) {
-            // custom type
-            resource = new UnknownBackingFileReadableResource((ReadableResource) tarPath);
-        } else {
-            tarFile = file(tarPath);
-            resource = new LocalResourceAdapter(new LocalFileStandInExternalResource(tarFile, fileSystem));
-        }
-        TarFileTree tarTree = new TarFileTree(tarFile, new MaybeCompressedFileResource(resource), getExpandDir(), fileSystem, directoryFileTreeFactory, streamHasher, fileHasher);
+        Provider<File> fileProvider = asFileProvider(tarPath);
+        Provider<ReadableResourceInternal> resource = providers.provider(() -> {
+            if (tarPath instanceof ReadableResourceInternal) {
+                return (ReadableResourceInternal) tarPath;
+            } else if (tarPath instanceof ReadableResource) {
+                // custom type
+                return new UnknownBackingFileReadableResource((ReadableResource) tarPath);
+            } else {
+                File tarFile = file(tarPath);
+                return new LocalResourceAdapter(new LocalFileStandInExternalResource(tarFile, fileSystem));
+            }
+        });
+
+        TarFileTree tarTree = new TarFileTree(fileProvider, resource.map(MaybeCompressedFileResource::new), getExpandDir(), fileSystem, directoryFileTreeFactory, streamHasher, fileHasher);
         return new FileTreeAdapter(tarTree, patternSetFactory);
     }
 
+    private Provider<File> asFileProvider(Object path) {
```

Comment:
Why do we need to convert the path to a provider at all? We could use the Object `path` here and then use a file resolver to convert the object to a path in `TarFileTree` adn `ZipFileTree`, right? Or, if the path is a provider, we could always use `provider.map(this::file)` in all cases, I don't think there is a need to check the type of the `Provider`.

Reply:
There are different questions in one. 

First, we need to convert the path to a `Provider` because this is how the `FileTreeAdapter` will track the dependencies internally. Before it was using a `File` which cannot track them.

Then the question is what is the supplied path? If it's a `ReadableResource`, there's special handling for the `TarFileTree` which needs to be taken care of.

If it's a `Provider`, then we have to perform the conversion at one point. There are 2 options:

1. like I've done here, so that the static type is `Provider<File>`
2. or defer the conversion to `FileResolver` so that it can convert a `RegularFile` to `File`

It felt like a less risky change to handle it only for this use case (1), since file resolvers are used in different contexts and supporting the conversion _in general_ might have unwanted side effects (like preventing the failure whenever you get a `RegularFile` but needed a `Directory`).